[FEATURE] Add support for IPF PTMs alignment

singjc opened the pull request:
Added `runType="DIA_IPF"` for alignment of PTMs IPF Proteomics results.

- Added sql queries for extracting data for IPF results from OSW file. [R/get_osw_query.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-7ecfaff8591177e9c2f32d9636ee1df7e06010557f7f440b991a649335c42d51) and [R/read_osw.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-34ddcdd1eeeba7c599e04e1d86c24dd16f0238d8e03e612d1de50337e250f726)
- Added extra params ([R/utils.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-60bef5af43a1bd541299ce9bceeb226afb23deaa5e637b145c85e05ffd552b0c))
    - `maxIPFFdrQuery` - A numeric value between 0 and 1. It is used to filter features from osw file which have SCORE_IPF.QVALUE less than itself. (For PTM IPF use)
    - `useIdentifying` - Set TRUE to use identifying transitions in alignment. (DEFAULT: FALSE)
- Added method `ipfReassignFDR` ([R/utils.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-60bef5af43a1bd541299ce9bceeb226afb23deaa5e637b145c85e05ffd552b0c)) to reassign FDR values to the aligned peaks. This assumes that the reference peak has good quality IPF FDRs, that had confident identifying transitions. This reassignment will only get done at the end of alignment, when the final results table is returned, and only if the `runType="DIA_IPF"`. The reassignment of the FDR is as follows:
    1.  If the MS2 FDR is less than the IPF FDR, then reassign the FDR using the MS2 FDR. If the feature was not identified by OpenSwath, then assign the user defined `alignedFDR`. (This excludes the reference run, the reference run's FDR is left as if)
    2. If the the assigned FDRs are greated than the user defined `alignedFDR`, then assign the user defined `alignedFDR`.
- Added tests to reflect the changes with `DIA_IPF`. [tests/testthat/test_align_dia_runs.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-c19d130fe1c4ca7b035ee4a20003bcfc05352b50b3d482dd20df799a28ce9ca4), [tests/testthat/test_get_osw_query.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-c08a227b5e0c49ec61cfcac6d8c5253aee19382fea786cd094a8d0e971bd86d5), [tests/testthat/test_read_osw.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-2b4b7974dd9f6871cfe318bf0a7f002cfeb5aee21d78a5df1141d3b576d82341) and [tests/testthat/test_utils.R](https://github.com/shubham1637/DIAlignR/pull/15/commits/b820a30d02f2dac15f13c842bd9b2415dc6bf3aa#diff-1c62f90a9dfb352c9cfd1b9163cd2995763a91108fc36274ec70f66c0f3e66fe)

shubham1637 commented on R/utils.R:
```diff
@@ -115,17 +121,31 @@ getMultipeptide <- function(precursors, features, applyFun=lapply, masters = NUL
+  if ( runType=="DIA_IPF" ){
+    data.table("transition_group_id" = analytes, "feature_id" = bit64::NA_integer64_,
+               "RT" = NA_real_, "intensity" = NA_real_, "leftWidth" = NA_real_,
+               "rightWidth" = NA_real_, "peak_group_rank" = NA_integer_, "ms2_m_score" = NA_real_, "m_score" = NA_real_)
```
Should we call `ms2_m_score` something like `IPF m-score`?

singjc replied:
`ms2_m_score` is not the `IPF m-score`, it's the `MS2 QVALUE` which is the default `m_score` that get returned by pyprophet. When doing **IPF**, then the `m_score` gets updated based on the **IPF** framework returning updated `QVALUE`s. I've named it `ms2_m_score` to retain the old `MS2 QVALUEs`. I would like to keep this name to keep it consistent with TRIC and Pyprophet?

shubham1637 replied:
Ok

shubham1637 commented on R/utils.R:
```diff
@@ -115,17 +121,31 @@ getMultipeptide <- function(precursors, features, applyFun=lapply, masters = NUL
+  if ( runType=="DIA_IPF" ){
+    data.table("transition_group_id" = rep(analytes, times = length(masters), each = 5L), "feature_id" = bit64::NA_integer64_,
+               "RT" = NA_real_, "intensity" = NA_real_, "leftWidth" = NA_real_,
+               "rightWidth" = NA_real_, "peak_group_rank" = NA_integer_, "ms2_m_score" = NA_real_, "m_score" = NA_real_,
```
Similar here

shubham1637 commented on R/utils.R:
```diff
@@ -86,10 +86,13 @@ getMultipeptide <- function(precursors, features, applyFun=lapply, masters = NUL
         df
       }), use.names=TRUE)
+      if ( runType=="DIA_IPF" ){
+        setcolorder(newdf, c('transition_group_id', 'feature_id', 'RT', 'intensity', 'leftWidth', 'rightWidth', 'peak_group_rank', 'm_score', 'run', 'alignment_rank', 'ms2_m_score'))
```
Add `setcolorder` in line 52

shubham1637 replied:
Not sure, why do we need to use `setcolorder`?

singjc replied:
> Add `setcolorder` in line 52

Just to make sure, which line is 52? You want me to add it to the `importFrom` statement? `#' @importFrom data.table rbindlist set setkeyv`

> Not sure, why do we need to use setcolorder?

I use `setcoloder` here to make sure the first 10 columns are `'transition_group_id', 'feature_id', 'RT', 'intensity', 'leftWidth', 'rightWidth', 'peak_group_rank', 'm_score', 'run', 'alignment_rank'`, otherwise this causes downstream issues when using the numbered indexes. If we don't `setcolorder`, then `alignment_rank` gets added to the very end (when there are other extra columns, i.e. `ms2_m_score`), and the rank updating will not be updating the right column anymore then.

shubham1637 replied:
Right. It would look like `@importFrom data.table rbindlist set setkeyv setcoloder`

shubham1637 replied:
Got it. Line 90 adds run and alignment_rank.

shubham1637 commented on R/utils.R:
```diff
@@ -526,6 +585,45 @@ getPrecursorSubset <- function(precursors, params){
+  ## Create Reference Run table with runName to map back to dt
+  runNameMap <- fileInfo$runName
+  names(runNameMap) <- rownames(fileInfo)
+  refRuns$ref_run <- dplyr::recode(refRuns$run, !!!runNameMap )
```
I suggest not to use dplyr because if the table is big 4-5 Gb, it will be really slow

singjc replied:
Make sense, resolved: https://github.com/shubham1637/DIAlignR/pull/15/commits/21c3b513bbbff86d55a1fdfb466daa4fab948f72

shubham1637 commented on R/read_osw.R:
```diff
@@ -363,15 +383,20 @@ fetchFeaturesFromRun <- function(filename, runID, maxFdrQuery = 1.00, runType =
 #' dim(features[[2]]) # 938  8
 #' @export
-getFeatures <- function(fileInfo, maxFdrQuery = 0.05, runType = "DIA_Proteomics", applyFun = lapply){
+getFeatures <- function(fileInfo, maxFdrQuery = 0.05, maxIPFFdrQuery = 0.05, runType = "DIA_Proteomics", applyFun = lapply){
```
Add definition of ms2_m_score in line 378

singjc replied:
Done

shubham1637 commented on R/utils.R:
```diff
@@ -526,6 +585,45 @@ getPrecursorSubset <- function(precursors, params){
+#' @keywords internal
+#' @seealso writeTables, getRefRun, getRunNames, paramsDIAlignR
+#' @importFrom dplyr recode select
+#' @import data.table
```
Add setnames here

singjc replied:
I import the entire package data.table `#' @import data.table`, should I still or change to importing individual things?

shubham1637 replied:
If you specify `@import data.table` then you need to use `data.table::Function` also `data.table::data.table` (if you create a new table).
If you specify `@importFrom data.table Function data.table` then you don't need to use `::`. You can use both `import` and `importFrom` in a function

singjc replied:
Ohh okay, cool makes sense, thanks
Done

shubham1637 commented on R/read_osw.R:
```diff
@@ -164,11 +165,19 @@ fetchPrecursorsInfo <- function(filename, runType = "DIA_Proteomics", selectIDs
+      DBI::dbDisconnect(con)})
+  } else {
+    precursorsInfo <- tryCatch(expr = { output <- DBI::dbSendQuery(con, statement = query)
+                       if(all && runType != "DIA_Metabolomics") {DBI::dbBind(output, list("CONTEXT"=context, "FDR"=maxPeptideFdr))}
```
I think we can just add another if statement here to accommodate `DIA_IPF` instead of repeating the lines?

singjc replied:
Done

shubham1637 commented on R/get_osw_query.R:
```diff
@@ -254,8 +316,33 @@ getPrecursorsQuery <- function(runType = "DIA_Proteomics", level = "Peptide")
+	    --TRANSITION.IDENTIFYING AS identifying
+      FROM PRECURSOR
+      INNER JOIN TRANSITION_PRECURSOR_MAPPING ON TRANSITION_PRECURSOR_MAPPING.PRECURSOR_ID = PRECURSOR.ID
+      INNER JOIN TRANSITION ON TRANSITION.ID = TRANSITION_PRECURSOR_MAPPING.TRANSITION_ID
```
It will be faster if you select IDENTIFYING and DETECTING columns before INNER JOIN

singjc replied:
So you mean before inner joining the `TRANISITION` table, to create a smaller sub-table to select the `IDENTIFYING` and `DETECTING`, and then use the smaller sub-table to inner join? Makes sense

shubham1637 replied:
Yes. It would matter here because TRANSITION table has a lot of columns.

shubham1637 commented on R/get_osw_query.R:
```diff
@@ -254,8 +316,33 @@ getPrecursorsQuery <- function(runType = "DIA_Proteomics", level = "Peptide")
+      WHERE PRECURSOR.DECOY = 0
+      AND (
+      TRANSITION.DETECTING=TRUE
+      OR TRANSITION.IDENTIFYING=$USE_IDENTIFYING --- #identifying
```
How does IPF calculate intensity? How many and which transitions does it use?

singjc replied:
> How does IPF calculate intensity?

IPF doesn't calculate intensity? It generates a set of theoretical identifying transitions, and then within the product m/z window it looks for overlapping or interfering signals that get identified and annotated. 

> How many and which transitions does it use?

It depends on the peptide, and the number of modifications, there can be as many as 100+ transitions. The IPF model itself uses all these identifying transitions.

I know this might be an issue with the alignment if there is not the same number of transitions being aligned, this part is a still a W.I.P.  I currently don't use the identifying transitions in the alignment, because they're not really needed for alignment, in my experience at least. What I have done previously for alignment (if wanting to use the identifying transitions), then I have taken the best scoring (based on posterior error probabilities) transitions, and just make sure it's an equal amount of transitions between the experiment and reference. I am still thinking of using a select few identifying transitions, but not entirely sure if it's necessary...

shubham1637 replied:
So only DETECTING transitions will be used to create a new feature if it is missing

shubham1637 commented on R/align_dia_runs.R:
```diff
@@ -129,10 +129,10 @@ alignTargetedRuns <- function(dataPath, outFile = "DIAlignR", params = paramsDIA
   message("Calculating global alignments.")
   start_time <- Sys.time()
-  globalFits <- getGlobalFits(refRuns, features, fileInfo, params[["globalAlignment"]],
+  globalFits <- DIAlignR:::getGlobalFits(refRuns, features, fileInfo, params[["globalAlignment"]],
```
No. Do not add `:::` colons. It produces warnings at Bioconductor biocCheck

singjc replied:
Right, sorry I used these for testing, forgot to remove it. Removed it

singjc reviewed (commented):
Made changes to address feedback

shubham1637 commented on R/align_dia_runs.R:
```diff
@@ -398,6 +404,7 @@ perBatch <- function(iBatch, peptides, multipeptide, refRuns, precursors, prec2c
     ref <- refRuns[rownum, "run"][[1]]
     idx <- (rownum - (iBatch-1)*batchSize)
     analytes <- analytesA[[idx]]
+    analytes <- analytesA[idx]
```
Why do we need this? Is it typo?

singjc replied:
Good catch, this is a typo from testing

shubham1637 commented on R/get_osw_query.R:
```diff
@@ -254,8 +316,36 @@ getPrecursorsQuery <- function(runType = "DIA_Proteomics", level = "Peptide")
+      PEPTIDE.MODIFIED_SEQUENCE AS sequence,
+      PRECURSOR.CHARGE AS charge,
+      PRECURSOR.GROUP_LABEL AS group_label
+      --TRANSITION.DETECTING AS detecting,
```
What does `--` do?

singjc replied:
The `--` in sql is to signify a comment, like `#` in R or `//` in C++

shubham1637 commented on R/utils.R:
```diff
@@ -526,6 +585,49 @@ getPrecursorSubset <- function(precursors, params){
+#' @keywords internal
+#' @seealso writeTables, getRefRun, getRunNames, paramsDIAlignR
+#' @import data.table
+#' @importFrom data.table merge.data.table setnames fifelse :=
```
I think the last should be ":=" or ':='
https://cran.r-project.org/web/packages/data.table/vignettes/datatable-importing.html

singjc replied:
good to know, added the quotes

shubham1637 commented:
When I run `devtoos::test()` on this branch. Few tests are failing.
[ FAIL 5 | WARN 0 | SKIP 5 | PASS 560 ]

singjc commented:
> When I run `devtoos::test()` on this branch. Few tests are failing.
> [ FAIL 5 | WARN 0 | SKIP 5 | PASS 560 ]

Okay I fixed all the errors that were coming up when running the tests.
1. Most of the errors came from other scripts that use `getFeatures`, because I added the `maxIPFFdrQuery` param, and in most cases where `getFeatures` is used, the params are set by position and not by it's named argument. I.e. `getFeatures(fileInfo, maxFdrQuery, params[["maxIPFFdrQuery"]], runType, lapply))`, I was wondering why don't we set input paramet values using the named argument, i.e,` getFeatures(fileInfo=fileInfo, maxFdrQuery=maxFdrQuery, maxIPFFdrQuery=params[["maxIPFFdrQuery"]], runType=runType, lapply=lapply)`? Is it just more efficient/faster to not explicitly assign?
2. Same errors occured with `getMultipeptide` because I added the `runType` param
3. I forgot to update the test for `test_get_osw_query`, because I changed where the `DETECTING` and `IDENTIFYING` transitions get selected in the query.

All tests seem to pass on my local copy now, let me know if there are still any failing tests on your tests.

```
══ Results ═══════════════════════════════════════════════════════════════════════════════════════════
Duration: 22.1 s

── Skipped tests  ────────────────────────────────────────────────────────────────────────────────────
• empty test (5)
• ropenms not available for testing. A conda environment with name TricEnvr is MUST for testing. (3)

[ FAIL 0 | WARN 0 | SKIP 8 | PASS 605 ]
```

shubham1637 commented:
Thanks. It is working on my end as well. I will test on Streptococcus data as well to see that output matches with previous version of it. Will merge it afterwards.

shubham1637 commented:
There is an error while doing BioConductor check due to large file size of the example data. One way to go about is to reduce the file size by having 10-20 peptides only. Or you can upload the data as data-package on Bioconductor separately and call that when you run your tests.
```
R CMD BiocCheck DIAlignR
* Checking package size...
    * ERROR: Package Source tarball exceeds Bioconductor size
      requirement.
        Package Size: 32.083 MB
        Size Requirement: 5.0000 MB
* Checking individual file sizes...
    * WARNING: The following files are over 5MB in size:
      'inst/ptms/xics/chludwig_K150309_004b_SW_1_16.chrom.sqMass
      inst/ptms/xics/chludwig_K150309_008_SW_1_4.chrom.sqMass
      inst/ptms/xics/chludwig_K150309_013_SW_0.chrom.sqMass'
```

shubham1637 commented on R/utils.R:
```diff
@@ -49,7 +49,7 @@ getRefRun <- function(peptideScores, applyFun=lapply){
+#' @importFrom data.table rbindlist set setkeyv setcolorder
 #' @importFrom bit64 NA_integer64_
 #' @inheritParams alignTargetedRuns
 #' @param precursors (data-frames) Contains precursors and associated transition IDs.
```
Add runType definition here. Warning from `devtools::check()`

singjc replied:
Done. I will make the example dataset smaller.

singjc commented:
@shubham1637 I shrunk the tests files.

shubham1637 commented:
Hi, I was thinking about it since last week but got a chance to try it out now.
```
(base) shubham@shubham-desktop:~/temp$ git clone git@github.com:shubham1637/DIAlignR.git
Cloning into 'DIAlignR'...
remote: Enumerating objects: 7226, done.
remote: Counting objects: 100% (994/994), done.
remote: Compressing objects: 100% (540/540), done.
remote: Total 7226 (delta 703), reused 679 (delta 452), pack-reused 6232
Receiving objects: 100% (7226/7226), 28.31 MiB | 3.13 MiB/s, done.
Resolving deltas: 100% (5357/5357), done.
Checking connectivity... done.
```

PR:
```
(base) shubham@shubham-desktop:~$ git clone git@github.com:singjc/DIAlignR.git
Cloning into 'DIAlignR'...
remote: Enumerating objects: 7690, done.
remote: Counting objects: 100% (1085/1085), done.
remote: Compressing objects: 100% (597/597), done.
remote: Total 7690 (delta 772), reused 740 (delta 484), pack-reused 6605
Receiving objects: 100% (7690/7690), 55.04 MiB | 3.72 MiB/s, done.
Resolving deltas: 100% (5726/5726), done.
Checking connectivity... done.
```
As you can see the original size of git repo is **28.31 MiB**. The PR size is **55.04 MiB**

```
(base) shubham@shubham-desktop:~/DIAlignR$ ll -h inst/ptms/xics/
total 736K
drwxrwxr-x 2 shubham shubham 4.0K Jul  5 18:05 ./
drwxrwxr-x 4 shubham shubham 4.0K Jul  5 18:05 ../
-rw-rw-r-- 1 shubham shubham 236K Jul  5 18:05 chludwig_K150309_004b_SW_1_16.chrom.sqMass
-rw-rw-r-- 1 shubham shubham 236K Jul  5 18:05 chludwig_K150309_008_SW_1_4.chrom.sqMass
-rw-rw-r-- 1 shubham shubham 256K Jul  5 18:05 chludwig_K150309_013_SW_0.chrom.sqMass

(base) shubham@shubham-desktop:~/DIAlignR$ git checkout 1db6d3dd4044b45549af542165ea16a75868076e
Note: checking out '1db6d3dd4044b45549af542165ea16a75868076e'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 1db6d3d... [ADD] Add runType definition to getMultiPeptide

(base) shubham@shubham-desktop:~/DIAlignR$ ll -h inst/ptms/xics/
total 33M
drwxrwxr-x 2 shubham shubham 4.0K Jul  5 18:07 ./
drwxrwxr-x 4 shubham shubham 4.0K Jul  5 18:05 ../
-rw-rw-r-- 1 shubham shubham  11M Jul  5 18:07 chludwig_K150309_004b_SW_1_16.chrom.sqMass
-rw-rw-r-- 1 shubham shubham  11M Jul  5 18:07 chludwig_K150309_008_SW_1_4.chrom.sqMass
-rw-rw-r-- 1 shubham shubham  11M Jul  5 18:07 chludwig_K150309_013_SW_0.chrom.sqMass
```

singjc commented:
Okay, solved.

```
git clone --branch feature/ipf_support git@github.com:singjc/DIAlignR.git 
Cloning into 'DIAlignR'...
remote: Enumerating objects: 8502, done.
remote: Counting objects: 100% (81/81), done.
remote: Compressing objects: 100% (34/34), done.
remote: Total 8502 (delta 50), reused 55 (delta 47), pack-reused 8421
Receiving objects: 100% (8502/8502), 27.01 MiB | 3.46 MiB/s, done.
Resolving deltas: 100% (6620/6620), done.
```

## Find large blobs specifically for old ptms example test files
```
 git rev-list --objects --all |
  git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' |
  sed -n 's/^blob //p' |
  sort --numeric-sort --key=2 |
  cut -c 1-12,41- |
  $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest | grep "inst/ptms/"
```

### Output
```
2e889c3dc310  236KiB inst/ptms/xics/chludwig_K150309_004b_SW_1_16.chrom.sqMass
6466577ead91  236KiB inst/ptms/xics/chludwig_K150309_008_SW_1_4.chrom.sqMass
4d94cab56620  256KiB inst/ptms/xics/chludwig_K150309_013_SW_0.chrom.sqMass
6c9658e1217b  760KiB inst/ptms/osw/merged.osw
18c8716a14eb  3.6MiB inst/ptms/osw/merged.osw
54e70e6a6df6   11MiB inst/ptms/xics/chludwig_K150309_004b_SW_1_16.chrom.sqMass
c2e99023ba25   11MiB inst/ptms/xics/chludwig_K150309_008_SW_1_4.chrom.sqMass
51f164323ec8   11MiB inst/ptms/xics/chludwig_K150309_013_SW_0.chrom.sqMass
```

## Get full hashes for objects to remove, and write them to a file rm_blobs
```
54e70e6a6df6ac5215338e23c0d7edbbe01d99ef
c2e99023ba25b72b5a2cc7c7ed31aa87509ce6f9
51f164323ec8820669c290f5bde110c24b43ec5a
18c8716a14ebe6b8ea7ba2a2025734a4bf5f838e
```

## Use git filter-repo to remove blobs
```
$ git filter-repo --strip-blobs-with-ids rm_blobs --force
Parsed 1025 commits
New history written in 0.46 seconds; now repacking/cleaning...
Repacking your repo and cleaning out old unneeded objects
HEAD is now at df0e577 [FIX] Shrink ptms test files
Enumerating objects: 7682, done.
Counting objects: 100% (7682/7682), done.
Delta compression using up to 12 threads
Compressing objects: 100% (1821/1821), done.
Writing objects: 100% (7682/7682), done.
Total 7682 (delta 5830), reused 7677 (delta 5828), pack-reused 0
Completely finished after 1.20 seconds.
```

## Recheck that blobs were removed
```
$ git rev-list --objects --all |
>   git cat-file --batch-check='%(objecttype) %(objectname) %(objectsize) %(rest)' |
>   sed -n 's/^blob //p' |
>   sort --numeric-sort --key=2 |
>   cut -c 1-12,41- |
>   $(command -v gnumfmt || echo numfmt) --field=2 --to=iec-i --suffix=B --padding=7 --round=nearest | grep "inst/ptms/"
2e889c3dc310  236KiB inst/ptms/xics/chludwig_K150309_004b_SW_1_16.chrom.sqMass
6466577ead91  236KiB inst/ptms/xics/chludwig_K150309_008_SW_1_4.chrom.sqMass
4d94cab56620  256KiB inst/ptms/xics/chludwig_K150309_013_SW_0.chrom.sqMass
6c9658e1217b  760KiB inst/ptms/osw/merged.osw
```
